Changes to run API server for Calico

caseydavenport opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

Still a WIP

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

caseydavenport commented on main.go:
```diff
@@ -155,7 +156,11 @@ func main() {
 	}
 
-	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+	// Disable deprecation warnings.
```
Not sure if we want this - but without it we seem to get a lot of warning logs.

caseydavenport replied:
I think it's be cause we use PSPs and they are now officially deprecated

caseydavenport commented on pkg/components/calico.go:
```diff
@@ -43,6 +43,11 @@ var (
+
+	ComponentCalicoAPIServer = component{
+		Version: "latest",
+		Image:   "caseydavenport/apiserver",
```
Need to update this before merge

tmjd replied:
Can this be replaced yet?

tmjd reviewed (commented):
Just a few comments I had made a little while back but apparently forgot to submit.

caseydavenport commented on pkg/controller/installation/core_controller.go:
```diff
@@ -790,6 +765,29 @@ func (r *ReconcileInstallation) Reconcile(ctx context.Context, request reconcile
+		if apiserver.Status.State != operator.TigeraStatusReady {
+			// API server is not yet ready. Wait for it.
+			r.status.SetDegraded("Waiting for API server to be ready", "")
+			return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
```
I was uncomfortable watching the APIServer because it would be a new dependency for this controller - i.e., a new version of the operator would fail unless the APIServer CRD was also present.

I suppose it's not the end of the world, but thought this would be more friendly / backwards compatible. I don't think we're saying the APIServer is required for OSS just yet?

caseydavenport commented on pkg/controller/apiserver/apiserver_controller.go:
```diff
@@ -180,18 +183,21 @@ func (r *ReconcileAPIServer) Reconcile(ctx context.Context, request reconcile.Re
-		r.status.SetDegraded(fmt.Sprintf("Waiting for network to be %s", operatorv1.TigeraSecureEnterprise), "")
-		return reconcile.Result{}, nil
-	}
+	variant := network.Variant
```
Yep. We can probably switch to using the status again actually.

The main reason I did this was back when I wanted the ability to run the APIServer _before_ calico/node, which meant the status field was unavailable.

I think even then we'd want to prefer status if available, then fall-back to spec. For now I'll just revert to status.

caseydavenport commented on pkg/render/apiserver.go:
```diff
@@ -344,24 +396,19 @@ func (c *apiServerComponent) apiServerServiceAccount() *corev1.ServiceAccount {
 	rules := []rbacv1.PolicyRule{
 		{
-			APIGroups: []string{
-				"extensions",
```
extensions is long-since deprecated, and all of the resources within it have been moved to other API groups (note this is different from apiextensions)

caseydavenport commented on pkg/render/apiserver.go:
```diff
@@ -114,35 +149,42 @@ func APIServer(k8sServiceEndpoint k8sapi.ServiceEndpoint,
+		if err != nil {
+			errMsgs = append(errMsgs, err.Error())
+		}
 	}
```
I think it's covered in the "else" clause?

tmjd replied:
:eyeglasses: I think I need some new glasses

caseydavenport commented on pkg/render/apiserver.go:
```diff
@@ -163,54 +205,97 @@ func (c *apiServerComponent) SupportedOSType() rmeta.OSType {
+	globalObjects := []client.Object{
+		c.calicoCustomResourcesClusterRole(),
+		c.calicoCustomResourcesClusterRoleBinding(),
+		c.delegateAuthClusterRoleBinding(),
```
Updated comment.

tmjd replied:
My main objection wasn't the comment really, but don't we want to clean up the non-in-place resources when switching between Calico/Enterprise? I was pointing out that the expectation that the majority of these are not in-place updates, and implicitly we need to handle them a bit differently.

caseydavenport commented on pkg/render/apiserver.go:
```diff
@@ -374,31 +421,54 @@ func (c *apiServerComponent) apiServiceAccountClusterRole() *rbacv1.ClusterRole
+			Verbs: []string{"*"},
+		},
+		{
+			// Calico Enterprise backing storage.
```
You're correct - looks like copy-paste error.

caseydavenport commented on pkg/controller/apiserver/apiserver_controller.go:
```diff
@@ -33,36 +31,33 @@ import (
 // Add creates a new APIServer Controller and adds it to the Manager. The Manager will set fields on the Controller
 // and Start it when the Manager is Started.
 func Add(mgr manager.Manager, opts options.AddOptions) error {
-	if !opts.EnterpriseCRDExists {
```
Yes.. very true. I guess which one do you think is preferrable? 

- Just add a new requirement for the APIServer CRD (we'll be including it by default anyway for new installs)
- Refactor to have three modes - OSS, OSS w/ APIServer, Enterprise - based on which CRs are present.

tmjd replied:
I think we should just make it required. We expect any updates applied will include the full set of CRDs so is there any reason to make it optional?
To not include the APIServer CRD would be a manual modification to the manifests that we prefer the users use unmodified.

caseydavenport commented:
@tmjd thanks for the review!

I think I've addressed most of the comments. I'm going to see if I can make the APIServer CRD optional "for real", unless you think it's just fine to make it mandatory moving forward?

tmjd reviewed (approved):
LGTM
